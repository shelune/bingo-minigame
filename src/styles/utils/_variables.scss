$black: #202020;
$white: #ffffff;
$orange: #ff6600;
$orange-dark: darken($orange, 10%);
$gray: #656565;
