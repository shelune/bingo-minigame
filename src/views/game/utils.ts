import { Tile } from "../../components/board/board";
import shuffle from "lodash.shuffle";

export function range(min: number, max: number) {
  return Array.from({ length: max - min + 1 }, (_, idx) => min + idx);
}

export function getNumberFromRange(range: number, excluding: number[]): number {
  const number = Math.round(Math.random() * (range - 1));
  return excluding.includes(number)
    ? getNumberFromRange(range, excluding)
    : number;
}

export function transpose(matrix: number[][]) {
  return matrix[0].map((_, colIdx) =>
    matrix.map((_, rowIdx) => matrix[rowIdx][colIdx])
  );
}

export function getBoardNumbers() {
  const numbers = [0, 1, 2, 3, 4].map((columnIdx) => {
    const selections = shuffle(range(columnIdx * 15 + 1, (columnIdx + 1) * 15));
    return [1, 2, 3, 4, 5].map((idx) => selections[idx]);
  });
  // transpose so the bingo card follows column not row
  const transposed = transpose(numbers);
  return transposed;
}

export function generateBoardTiles(numbers: number[]) {
  const tiles: Tile[] = numbers.map((number, idx) => {
    return {
      id: idx,
      value: number,
      marked: false,
    };
  });
  return tiles;
}

export function checkRow(board: Tile[], idxOnBoard: number) {
  const rowIdx = Math.floor(idxOnBoard / 5);
  const markedCount = [0, 1, 2, 3, 4].reduce((accu, curr) => {
    const boardIdx = rowIdx * 5 + curr;
    if (board[boardIdx].marked) {
      return accu + 1;
    }
    return accu;
  }, 0);
  return markedCount === 5;
}

export function checkColumn(board: Tile[], idxOnBoard: number) {
  const columnIdx = idxOnBoard % 5;
  const markedCount = [0, 1, 2, 3, 4].reduce((accu, curr) => {
    const boardIdx = curr * 5 + columnIdx;
    if (board[boardIdx].marked) {
      return accu + 1;
    }
    return accu;
  }, 0);
  return markedCount === 5;
}

export function checkDiagonal(board: Tile[]) {
  const diagonalForwardTiles = board.filter((tile) => {
    return tile.id > 0 && tile.id % 4 === 0 && tile.id < 24;
  });
  const diagonalBackwardTiles = board.filter((tile) => {
    return tile.id % 6 === 0;
  });
  return (
    diagonalForwardTiles.every((tile) => tile.marked) ||
    diagonalBackwardTiles.every((tile) => tile.marked)
  );
}
