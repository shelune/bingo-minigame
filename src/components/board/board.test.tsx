import React from "react";
import { screen, render } from "@testing-library/react";

import Board from "./board";
import { mockBareBoard } from "../../views/game/utils.mock";

describe("Board", () => {
  it("should render", () => {
    const { container } = render(
      <Board tiles={mockBareBoard} onMark={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should call onMark when clicking a tile", () => {
    const mockOnMark = jest.fn();
    render(<Board tiles={mockBareBoard} onMark={mockOnMark} />);
    screen.getAllByTestId("tile")[0].click();
    expect(mockOnMark).toHaveBeenCalled();
  });
});
