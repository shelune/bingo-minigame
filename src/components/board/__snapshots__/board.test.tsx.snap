// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render 1`] = `
<div>
  <div
    class="content"
  >
    <h1
      class="title"
    >
      <div>
        <span>
          B
        </span>
      </div>
      <div>
        <span>
          I
        </span>
      </div>
      <div>
        <span>
          N
        </span>
      </div>
      <div>
        <span>
          G
        </span>
      </div>
      <div>
        <span>
          O
        </span>
      </div>
    </h1>
    <div
      class="board"
    >
      <div
        class="tile"
        data-testid="tile"
      >
        2
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        16
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        39
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        55
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        61
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        9
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        25
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        32
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        53
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        67
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        10
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        19
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        Free
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        59
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        63
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        13
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        27
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        43
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        47
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        64
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        7
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        18
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        38
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        58
      </div>
      <div
        class="tile"
        data-testid="tile"
      >
        71
      </div>
    </div>
  </div>
</div>
`;
